lib/charms/operator_libs_linux/v0/grub.py

```diff
@@ -29,4 +70,288 @@
 # to 0 if you are raising the major API version
 LIBPATCH = 1
 
-# TODO: add your code here! Happy coding!
+GRUB_DIRECTORY = Path("/etc/default/grub.d/")
+CHARM_CONFIG_PREFIX = "90-juju"
+GRUB_CONFIG = GRUB_DIRECTORY / "95-juju-charm.cfg"
+CONFIG_HEADER = f"""# This config file was produced by GRUB lib v{LIBAPI}.{LIBPATCH}.
+# https://charmhub.io/operator-libs-linux/libraries/grub
+"""
+FILE_LINE_IN_DESCRIPTION = "#   {path}"
+CONFIG_DESCRIPTION = """
+# This file represents the output of the GRUB lib, which can combine multiple
+# configurations into a single file like this.
+#
+# Original files:
+{configs}
+#
+# If you change this file, run 'update-grub' afterwards to update
+# /boot/grub/grub.cfg.
+# For full documentation of the options in this file, see:
+#   info -f grub -n 'Simple configuration'
+"""
+
+# TODO: use shlex
+
+
+class ValidationError(ValueError):
+    """Exception representing value validation error."""
+
+    def __init__(self, key: str, message: str) -> None:
+        self.key = key
+        self.message = message
+        super().__init__(message)
+
+    def __str__(self) -> str:
+        """Return string representation of error."""
+        return self.message
+
+
+class IsContainerError(Exception):
+    """Exception if local machine is container."""
+
+
+class ApplyError(Exception):
+    """Exception if applying new config failed."""
+
+
+def _parse_config(stream: io.TextIOWrapper) -> Dict[str, str]:
+    """Parse config file lines."""
+    config = {}
+    for line in stream:
+        line = line.strip()
+        if not line or line.startswith("#"):
+            logger.debug("skipping line `%s`", line)
+            continue
+
+        key, value = line.split("=", 1)
+        if key in config:
+            logger.warning("key %s is duplicated in config", key)
+
+        config[key] = value
+
+    return config
+
+
+def _load_config(path: Path) -> Dict[str, str]:
+    """Load config file from /etc/default/grub.d/ directory."""
+    if not path.exists():
+        raise FileNotFoundError("GRUB config file %s was not found", path)
+
+    with open(path, "r", encoding="UTF-8") as file:
+        config = _parse_config(file)
+
+    logger.info("GRUB config file %s was loaded", path)
+    logger.debug("config file %s", config)
+    return config
+
+
+def _save_config(path: Path, config: Dict[str, str], header: str = CONFIG_HEADER) -> None:
+    """Save GRUB config file."""
+    if path.exists():
+        logger.debug("GRUB config %s already exist and it will overwritten", path)
+
+    context = [f"{key}={value}" for key, value in config.items()]
+    with open(path, "w", encoding="UTF-8") as file:
+        file.writelines([header, *context])
+
+    logger.info("GRUB config file %s was saved", path)
+
+
+def check_update_grub() -> bool:
+    """Check if an update to /boot/grub/grub.cfg is available."""
```

Comment:
For boolean-returning functions, I like the word "Report whether an update..." as that indicates what the return value is. Similar for `is_container` below, "Report whether the local machine is a container."""